src/prg_response_mod.F90

```diff
@@ -831,6 +832,107 @@ subroutine prg_project_response(rsp_bml,over_bml,spindex,norbi,coordinates,rspfu
 
   end subroutine prg_project_response
 
+
+  !>  First-order Canonical Density Matrix Perturbation Theory.
+  !! \brief  Assuming known mu0 and representation in H0's eigenbasis Q,
+  !! where H0 and P0 become diagonal.
+  !! (mu0, eigenvalues e and eigenvectors Q of H0 are assumed known)
+  !! Based on PRL 92, 193001 (2004) and PRE 92, 063301 (2015).
+  !! \param P1_bml First-order canonical response output.
+  !! \param H1_bml Perturbative hamiltonian input.
+  !! \param Nocc Number of ocupied orbitals.
+  !! \param beta Inverse electronic temperature.
+  !! \param evals Eigenvalues of the system.
+  !! \param mu0 Chemical potential.
+  !! \param m Number of recursive steps.
+  !! \param HDIM Number of orbitals - Hamiltonina zise.
+  subroutine  prg_canon_response_vector(P1_bml,H1_bml,Nocc,beta,evals,mu0,m,thresh,HDIM)
+
+    real(dp), parameter   :: ONE = 1.D0, TWO = 2.D0, ZERO = 0.D0
+    type(bml_matrix_t), intent(inout)    ::  P1_bml, H1_bml
+    type(bml_matrix_t)    :: X_BML
+    integer, intent(in)   :: HDIM, m ! Nocc = Number of occupied orbitals, m= Number of recursion steps
+    real(dp), intent(in)  :: evals(HDIM), thresh !Q and e are eigenvectors and eigenvalues of H0
+    real(dp), intent(in)  :: beta, mu0 ! Electronic temperature and chemical potential
+    real(dp), intent(in)  :: nocc
+    real(dp)              :: X(HDIM,HDIM), DX1(HDIM,HDIM), Y(HDIM,HDIM) !Temporary matrices
+    real(dp)              :: h_0(HDIM), p_0(HDIM), dPdmu(HDIM), p_02(HDIM),iD0(HDIM)
+    real(dp)              :: cnst, kB, mu1
+    real(dp), allocatable :: FOCC(:), row1(:)
+    real(dp)              :: tmp, tmp2, dmu
+    integer               :: i, j, k
+    character(20) :: bml_type
+
+    kB = 8.61739e-5        ! (eV/K)
+
+    ALLOCATE(FOCC(HDIM))
```

Comment:
Please convert Fortran keywords to lower case, i.e.

```suggestion
    allocate(FOCC(HDIM))
```

Reply:
Ok, thanks. Just found there is new merge, need to rebase as well.